File: Forms/MainForm.vb
Public Class MainForm
    Private WithEvents _er As New GIGATMS.Devices.ER750.ER750Lib

    Private Delegate Sub UpdateUxDeviceListViewCallBack(ByVal deviceStatus As GIGATMS.Devices.ER750.Parameters.DeviceStatusFormat)
    Private Delegate Sub UpdateUxEventDataListViewCallBack(ByVal deviceIP As String, _
                                                           ByVal deviceStatus As GIGATMS.Devices.ER750.Parameters.EventDataFormat)
    Private Delegate Sub UpdateUxEventMessageListViewCallBack(ByVal source As String, _
                                                              ByVal message As String)

    Private Sub RefreshDevices()
        uxDeviceListView.Items.Clear()
        _er.Broadcast(uxLocalIps.Text)
    End Sub

    Private Sub Button1_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Button1.Click
        RefreshDevices()
    End Sub

    Private Sub UpdateUxEventMessageListView(ByVal source As String, _
                                             ByVal message As String)

        Dim li As Windows.Forms.ListViewItem

        Try
            li = uxEventMessageListView.Items.Add(Now)
            li.SubItems.Add(source)
            li.SubItems.Add(message)
            li.EnsureVisible()
        Catch ex As Exception
            MsgBox(ex.Message)
        End Try
    End Sub

    Private Sub _er_ErrorOccured(ByVal source As String, ByVal errorMessage As String) Handles _er.ErrorOccured
        If Me.InvokeRequired() Then
            Dim cb As New UpdateUxEventMessageListViewCallBack(AddressOf UpdateUxEventMessageListView)
            Me.Invoke(cb, New Object() {source, "***Error***: " & errorMessage})
        Else
            UpdateUxEventMessageListView(source, "***Error***: " & errorMessage)
        End If
    End Sub

    Private Sub _er_EventMessage(ByVal source As String, ByVal eventMessage As String) Handles _er.EventMessage
        If Me.InvokeRequired() Then
            Dim cb As New UpdateUxEventMessageListViewCallBack(AddressOf UpdateUxEventMessageListView)
            Me.Invoke(cb, New Object() {source, eventMessage})
        Else
            UpdateUxEventmessageListView(source, eventMessage)
        End If
    End Sub


    Private Sub _er_ReceivedData(ByVal remoteIpAddress As String, ByVal remotePort As Integer, ByVal dataByteArray() As Byte) Handles _er.ReceivedData
        Debug.Print(remoteIpAddress)
    End Sub

    Private Sub UpdateUxDeviceListView(ByVal deviceStatus As GIGATMS.Devices.ER750.Parameters.DeviceStatusFormat)
        Dim li As Windows.Forms.ListViewItem

        Try
            li = uxDeviceListView.Items.Add(deviceStatus.IpAddress)
            li.SubItems.Add(deviceStatus.MacAddress)
            li.SubItems.Add(deviceStatus.FirmwareVersion)
            li.SubItems.Add(deviceStatus.DeviceName)
            li.EnsureVisible()
        Catch ex As Exception
            MsgBox(ex.Message)
        End Try

    End Sub


    Private Sub UpdateUxEventDataListView(ByVal deviceIP As String, _
                                          ByVal eventData As GIGATMS.Devices.ER750.Parameters.EventDataFormat)
        Dim li As Windows.Forms.ListViewItem

        Try
            li = uxEventDataListView.Items.Add(deviceIP)
            li.SubItems.Add(Now)
            li.SubItems.Add(eventData.DataHexString)
            li.SubItems.Add(eventData.DeviceName)
            li.EnsureVisible()

            RunBatchCommand(deviceIP)
        Catch ex As Exception
            MsgBox(ex.Message)
        End Try
    End Sub

    Private Sub _er_ReceivedDeviceStatus(ByRef deviceStatus As GIGATMS.Devices.ER750.Parameters.DeviceStatusFormat) Handles _er.ReceivedDeviceStatus
        If Me.InvokeRequired() Then
            Dim cb As New UpdateUxDeviceListViewCallBack(AddressOf UpdateUxDeviceListView)
            Me.Invoke(cb, deviceStatus)
        Else
            UpdateUxDeviceListView(deviceStatus)
        End If
    End Sub

    Private Sub MainForm_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
        Me.Text = My.Application.Info.Title & _
                  " V" & My.Application.Info.Version.Major & "." & _
                  My.Application.Info.Version.Minor & "R" & _
                  My.Application.Info.Version.Revision

        uxFormInitializeTimer.Enabled = True

    End Sub


    Private Sub InitializeUx()

        With uxDeviceListView
            .View = View.Details
            .FullRowSelect = True
            .GridLines = True
            .HideSelection = False
            .Columns.Clear()
            .Columns.Add("IP", CInt(.Width / 5))
            .Columns.Add("MAC", CInt(.Width / 4))
            .Columns.Add("Version", CInt(.Width / 7))
            .Columns.Add("Name", CInt(.Width / 2))
        End With

        Dim localIPs() As String = Nothing
        Dim localIpCount As Integer
        Dim i As Integer

        _er.GetLocalIP(localIpCount, localIPs)
        For i = 0 To localIpCount - 1
            uxLocalIps.Items.Add(localIPs(i))
        Next

        If localIpCount > 0 Then
            uxLocalIps.SelectedIndex = 0
        End If

        With uxEventDataListView
            .View = View.Details
            .FullRowSelect = True
            .GridLines = True
            .HideSelection = False
            .Columns.Clear()
            .Columns.Add("IP", CInt(.Width / 5))
            .Columns.Add("Date/Time", CInt(.Width / 3))
            .Columns.Add("Event Data", CInt(.Width / 5))
            .Columns.Add("Device Name", CInt(.Width / 2))
        End With

        With uxEventMessageListView
            .View = View.Details
            .FullRowSelect = True
            .GridLines = True
            .HideSelection = False
            .Columns.Clear()
            .Columns.Add("Date/Time", CInt(.Width / 4))
            .Columns.Add("Sourece", CInt(.Width / 4))
            .Columns.Add("Message", CInt(.Width / 2))
        End With


        With uxLedBuzzerControlSelectionList
            .DropDownStyle = ComboBoxStyle.DropDownList
            .Items.Clear()
            .Items.Add("0- Turn LED off")
            .Items.Add("1- Green on")
            .Items.Add("2- Green off")
            .Items.Add("3- Red on")
            .Items.Add("4- Red off")
            .Items.Add("5- 1 beep")
            .Items.Add("6- 3 beeps")
            .Items.Add("7- Green on with 1 beep")
            .Items.Add("8- Red on with 3 beeps")
            .Items.Add("9- Red, Green LED on")
            .Items.Add("C- Buzzer always beep")
            .Items.Add("D- Buzzer off")
            .SelectedIndex = 0
        End With
    End Sub

    Private Sub uxFormInitializeTimer_Tick(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles uxFormInitializeTimer.Tick
        uxFormInitializeTimer.Enabled = False

        InitializeUx()
    End Sub

    Private Sub uxDeviceListView_SelectedIndexChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles uxDeviceListView.SelectedIndexChanged
        If uxDeviceListView.SelectedItems.Count > 0 Then
            uxSelectedDeviceIp.Text = uxDeviceListView.SelectedItems.Item(0).Text
        End If
    End Sub

    Private Sub Button4_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Button4.Click
        End
    End Sub

    Private Sub Button2_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles uxStartListen.Click
        Select Case uxStartListen.Text
            Case "Start Listening"
                _er.StartListen(uxLocalIps.Text, uxListenPort.Value)
                uxStartListen.Text = "Stop Listening"
            Case "Stop Listening"
                _er.StopListen()
                uxStartListen.Text = "Start Listening"
            Case Else
                Stop
        End Select


    End Sub

    Private Sub _er_ReceivedEventData(ByVal deviceIP As String, _
                                      ByRef eventData As GIGATMS.Devices.ER750.Parameters.EventDataFormat) Handles _er.ReceivedEventData
        If Me.InvokeRequired() Then
            Dim cb As New UpdateUxEventDataListViewCallBack(AddressOf UpdateUxEventDataListView)
            Me.Invoke(cb, New Object() {deviceIP, eventData})
        Else
            UpdateUxEventDataListView(deviceIP, eventData)
        End If
    End Sub

    Private Sub uxEventDataListView_SelectedIndexChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles uxEventDataListView.SelectedIndexChanged
        If uxEventDataListView.SelectedItems.Count > 0 Then
            uxSelectedDeviceIp.Text = uxEventDataListView.SelectedItems.Item(0).Text
        End If
    End Sub

    Private Sub Button3_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Button3.Click
        Try
            _er.OpenDoor(uxSelectedDeviceIp.Text, _
                         uxRemotePort.Value, _
                         uxOpenDoorDuration.Value)
        Catch ex As Exception
            MsgBox(ex.Message)
        End Try
    End Sub

    Private Sub Button6_Click_1(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Button6.Click
        uxEventDataListView.Items.Clear()
    End Sub

    Private Sub Button2_Click_1(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Button2.Click
        uxEventMessageListView.Items.Clear()
    End Sub

    Private Sub uxSelectedDeviceIp_TextChanged(sender As Object, e As EventArgs) Handles uxSelectedDeviceIp.TextChanged

    End Sub

    Private Sub GroupBox4_Enter(sender As Object, e As EventArgs) Handles GroupBox4.Enter

    End Sub

    Private Sub uxOpenDoorDuration_ValueChanged(sender As Object, e As EventArgs) Handles uxOpenDoorDuration.ValueChanged

    End Sub

    Private Sub GroupBox3_Enter(sender As Object, e As EventArgs) Handles GroupBox3.Enter

    End Sub

    Private Sub Button5_Click(sender As Object, e As EventArgs) Handles Button5.Click
        Try
            Dim selectedFunction As Integer

            Select Case uxLedBuzzerControlSelectionList.Text.Substring(0, 1)
                Case "C"
                    selectedFunction = Asc("C") - Asc("0")
                Case "D"
                    selectedFunction = Asc("D") - Asc("0")
                Case Else
                    selectedFunction = uxLedBuzzerControlSelectionList.SelectedIndex
            End Select

            _er.ControlLedBuzzer(uxSelectedDeviceIp.Text,
                                 uxRemotePort.Value,
                                 selectedFunction)
        Catch ex As Exception
            MsgBox(ex.Message)
        End Try
    End Sub

    Private Sub uxLocalIps_SelectedIndexChanged(sender As Object, e As EventArgs) Handles uxLocalIps.SelectedIndexChanged

    End Sub

    Private Sub TextBox1_TextChanged(sender As Object, e As EventArgs) Handles TextBox1.TextChanged

    End Sub

    Private Sub RunBatchCommand(ByVal deviceIP As String)
        If deviceIP <> "" Then
            uxOpenDoorTimer.Tag = deviceIP
            uxOpenDoorTimer.Enabled = True
        Else
            MsgBox("Please select a device to run Batch Command")
        End If

    End Sub

    Private Sub Button7_Click(sender As Object, e As EventArgs) Handles Button7.Click

        RunBatchCommand(uxSelectedDeviceIp.Text)
    End Sub

    Private Sub uxBatchCommandTimer_Tick(sender As Object, e As EventArgs) Handles uxOpenDoorTimer.Tick
        uxOpenDoorTimer.Enabled = False

        Dim deviceIP As String

        deviceIP = uxOpenDoorTimer.Tag

        _er.OpenDoor(deviceIP,
                     uxRemotePort.Value,
                     uxOpenDoorDuration.Value,
                     False)
        uxControlLedBuzzerTimer.Tag = deviceIP

        uxControlLedBuzzerTimer.Enabled = True
    End Sub

    Private Sub uxControlLedBuzzerTimer_Tick(sender As Object, e As EventArgs) Handles uxControlLedBuzzerTimer.Tick
        uxControlLedBuzzerTimer.Enabled = False

        Dim deviceIP As String

        deviceIP = uxControlLedBuzzerTimer.Tag

        _er.ControlLedBuzzer(deviceIP,
                             uxRemotePort.Value,
                             uxLedBuzzerControlSelectionList.SelectedIndex,
                             False)
    End Sub
End Class

File: src/MainForm.cs
using System;
using System.Diagnostics;
using System.IO;
using System.Windows.Forms;
using System.Threading.Tasks;
using System.Reflection;

namespace RER750Manager
{
    public partial class MainForm : Form
    {
        private GIGATMS.Devices.ER750.ER750Lib _er = new GIGATMS.Devices.ER750.ER750Lib();
        private readonly string _debugLogPath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.Desktop), "ER750_Debug.log");

        // Delegates
        private delegate void UpdateUxDeviceListViewCallBack(GIGATMS.Devices.ER750.Parameters.DeviceStatusFormat deviceStatus);
        private delegate void UpdateUxEventDataListViewCallBack(string deviceIP, GIGATMS.Devices.ER750.Parameters.EventDataFormat deviceStatus);
        private delegate void UpdateUxEventMessageListViewCallBack(string source, string message);

        public MainForm()
        {
            InitializeComponent();
            RegisterEventHandlers();
            LogDebug("Application initialized");
        }

        #region Debugging Utilities
        private void LogDebug(string message)
        {
            try
            {
                string timestamp = DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss.fff");
                string logMessage = $"[{timestamp}] {message}{Environment.NewLine}";
                File.AppendAllText(_debugLogPath, logMessage);
                Debug.WriteLine(logMessage);
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"Logging failed: {ex.Message}");
            }
        }

        private void DumpObjectToLog(object obj, string prefix = "")
        {
            try
            {
                if (obj == null)
                {
                    LogDebug($"{prefix}Object is null");
                    return;
                }

                foreach (var prop in obj.GetType().GetProperties())
                {
                    try
                    {
                        LogDebug($"{prefix}{prop.Name} = {prop.GetValue(obj, null)}");
                    }
                    catch (Exception ex)
                    {
                        LogDebug($"{prefix}Failed to read {prop.Name}: {ex.Message}");
                    }
                }
            }
            catch (Exception ex)
            {
                LogDebug($"{prefix}Dump failed: {ex.Message}");
            }
        }
        #endregion

        #region Event Handler Registration
        private void RegisterEventHandlers()
        {
            try
            {
                _er.ReceivedDeviceStatus += _er_ReceivedDeviceStatus;
                _er.ReceivedEventData += _er_ReceivedEventData;
                _er.EventMessage += _er_EventMessage;
                _er.ErrorOccured += _er_ErrorOccured;
                _er.ReceivedData += _er_ReceivedData;
                LogDebug("Event handlers registered successfully");
            }
            catch (Exception ex)
            {
                LogDebug($"Failed to register event handlers: {ex}");
                MessageBox.Show($"Failed to initialize communication: {ex.Message}");
            }
        }
        #endregion

        #region Core Device Communication
        private async void RefreshDevices()
        {
            try
            {
                LogDebug($"Starting broadcast using IP: {uxLocalIps.Text}");
                uxDeviceListView.BeginUpdate();
                uxDeviceListView.Items.Clear();

                if (string.IsNullOrEmpty(uxLocalIps.Text))
                {
                    MessageBox.Show("Please select a local IP first");
                    return;
                }

                // Remove the bool assignment since Broadcast() returns void
                _er.Broadcast(uxLocalIps.Text);
                LogDebug("Broadcast initiated");

                await Task.Delay(1000);
                LogDebug($"Broadcast completed. Found {uxDeviceListView.Items.Count} devices");
            }
            catch (Exception ex)
            {
                LogDebug($"Broadcast error: {ex}");
                MessageBox.Show($"Broadcast failed: {ex.Message}");
            }
            finally
            {
                uxDeviceListView.EndUpdate();
            }
        }
        private void RunBatchCommand(string deviceIP)
        {
            try
            {
                if (string.IsNullOrEmpty(deviceIP))
                {
                    LogDebug("Batch command failed - no device IP specified");
                    MessageBox.Show("Please select a device to run Batch Command");
                    return;
                }

                LogDebug($"Starting batch command for {deviceIP}");
                uxOpenDoorTimer.Tag = deviceIP;
                uxOpenDoorTimer.Enabled = true;
            }
            catch (Exception ex)
            {
                LogDebug($"Batch command failed: {ex}");
                MessageBox.Show(ex.Message);
            }
        }
        #endregion

        #region UI Update Methods
        private void UpdateUxDeviceListView(GIGATMS.Devices.ER750.Parameters.DeviceStatusFormat deviceStatus)
        {
            try
            {
                LogDebug($"Updating UI with device: {deviceStatus.IpAddress}");
                
                var li = new ListViewItem(deviceStatus.IpAddress);
                li.SubItems.Add(deviceStatus.MacAddress ?? "N/A");
                li.SubItems.Add(deviceStatus.FirmwareVersion ?? "N/A");
                li.SubItems.Add(deviceStatus.DeviceName ?? "N/A");

                if (uxDeviceListView.InvokeRequired)
                {
                    uxDeviceListView.Invoke(new Action(() => {
                        uxDeviceListView.Items.Add(li);
                        li.EnsureVisible();
                    }));
                }
                else
                {
                    uxDeviceListView.Items.Add(li);
                    li.EnsureVisible();
                }
            }
            catch (Exception ex)
            {
                LogDebug($"Device list update failed: {ex}");
            }
        }

        private void UpdateUxEventDataListView(string deviceIP, GIGATMS.Devices.ER750.Parameters.EventDataFormat eventData)
        {
            try
            {
                LogDebug($"New event data from {deviceIP}");
                
                var li = new ListViewItem(deviceIP);
                li.SubItems.Add(DateTime.Now.ToString());
                li.SubItems.Add(eventData.DataHexString ?? "N/A");
                li.SubItems.Add(eventData.DeviceName ?? "N/A");

                if (uxEventDataListView.InvokeRequired)
                {
                    uxEventDataListView.Invoke(new Action(() => {
                        uxEventDataListView.Items.Add(li);
                        li.EnsureVisible();
                    }));
                }
                else
                {
                    uxEventDataListView.Items.Add(li);
                    li.EnsureVisible();
                }

                RunBatchCommand(deviceIP);
            }
            catch (Exception ex)
            {
                LogDebug($"Event data update failed: {ex}");
            }
        }

        private void UpdateUxEventMessageListView(string source, string message)
        {
            try
            {
                var li = new ListViewItem(DateTime.Now.ToString());
                li.SubItems.Add(source ?? "N/A");
                li.SubItems.Add(message ?? "N/A");

                if (uxEventMessageListView.InvokeRequired)
                {
                    uxEventMessageListView.Invoke(new Action(() => {
                        uxEventMessageListView.Items.Add(li);
                        li.EnsureVisible();
                    }));
                }
                else
                {
                    uxEventMessageListView.Items.Add(li);
                    li.EnsureVisible();
                }
            }
            catch (Exception ex)
            {
                LogDebug($"Message update failed: {ex}");
            }
        }
        #endregion

        #region ER750 Library Event Handlers
        private void _er_ReceivedDeviceStatus(ref GIGATMS.Devices.ER750.Parameters.DeviceStatusFormat deviceStatus)
        {
            LogDebug($"Device status received from {deviceStatus.IpAddress}");
            DumpObjectToLog(deviceStatus, "DeviceStatus_");
            
            var statusCopy = deviceStatus;
            UpdateUxDeviceListView(statusCopy);
        }

        private void _er_ReceivedEventData(string deviceIP, ref GIGATMS.Devices.ER750.Parameters.EventDataFormat eventData)
        {
            LogDebug($"Event data received from {deviceIP}");
            DumpObjectToLog(eventData, "EventData_");
            
            var dataCopy = eventData;
            UpdateUxEventDataListView(deviceIP, dataCopy);
        }

        private void _er_EventMessage(string source, string eventMessage)
        {
            LogDebug($"Event message: {source} - {eventMessage}");
            UpdateUxEventMessageListView(source, eventMessage);
        }

        private void _er_ErrorOccured(string source, string errorMessage)
        {
            LogDebug($"ERROR [{source}]: {errorMessage}");
            UpdateUxEventMessageListView(source, $"***ERROR***: {errorMessage}");
        }

        private void _er_ReceivedData(string remoteIpAddress, int remotePort, byte[] dataByteArray)
        {
            string hexData = dataByteArray != null ? BitConverter.ToString(dataByteArray) : "null";
            LogDebug($"Raw data from {remoteIpAddress}:{remotePort} - {hexData}");
        }
        #endregion

        #region Form Events
        private void MainForm_Load(object sender, EventArgs e)
        {
            try
            {
                var version = Assembly.GetExecutingAssembly().GetName().Version;
                this.Text = $"{Application.ProductName} V{version.Major}.{version.Minor}R{version.Revision}";
                uxFormInitializeTimer.Enabled = true;
                LogDebug($"MainForm loaded. Version: {version}");
            }
            catch (Exception ex)
            {
                LogDebug($"Load error: {ex}");
                this.Text = Application.ProductName;
            }
        }

        private void uxFormInitializeTimer_Tick(object sender, EventArgs e)
        {
            uxFormInitializeTimer.Enabled = false;
            try
            {
                InitializeUx();
                LogDebug("UI initialization completed");
            }
            catch (Exception ex)
            {
                LogDebug($"UI init failed: {ex}");
            }
        }

        private void InitializeUx()
        {
            uxDeviceListView.View = View.Details;
            uxDeviceListView.FullRowSelect = true;
            uxDeviceListView.GridLines = true;
            uxDeviceListView.HideSelection = false;
            uxDeviceListView.Columns.Clear();
            uxDeviceListView.Columns.Add("IP", uxDeviceListView.Width / 5);
            uxDeviceListView.Columns.Add("MAC", uxDeviceListView.Width / 4);
            uxDeviceListView.Columns.Add("Version", uxDeviceListView.Width / 7);
            uxDeviceListView.Columns.Add("Name", uxDeviceListView.Width / 2);

            string[] localIPs = null;
            int localIpCount = 0;
            
            _er.GetLocalIP(ref localIpCount, ref localIPs);
            for (int i = 0; i < localIpCount; i++)
            {
                uxLocalIps.Items.Add(localIPs[i]);
            }

            if (localIpCount > 0)
            {
                uxLocalIps.SelectedIndex = 0;
            }

            uxEventDataListView.View = View.Details;
            uxEventDataListView.FullRowSelect = true;
            uxEventDataListView.GridLines = true;
            uxEventDataListView.HideSelection = false;
            uxEventDataListView.Columns.Clear();
            uxEventDataListView.Columns.Add("IP", uxEventDataListView.Width / 5);
            uxEventDataListView.Columns.Add("Date/Time", uxEventDataListView.Width / 3);
            uxEventDataListView.Columns.Add("Event Data", uxEventDataListView.Width / 5);
            uxEventDataListView.Columns.Add("Device Name", uxEventDataListView.Width / 2);

            uxEventMessageListView.View = View.Details;
            uxEventMessageListView.FullRowSelect = true;
            uxEventMessageListView.GridLines = true;
            uxEventMessageListView.HideSelection = false;
            uxEventMessageListView.Columns.Clear();
            uxEventMessageListView.Columns.Add("Date/Time", uxEventMessageListView.Width / 4);
            uxEventMessageListView.Columns.Add("Source", uxEventMessageListView.Width / 4);
            uxEventMessageListView.Columns.Add("Message", uxEventMessageListView.Width / 2);

            uxLedBuzzerControlSelectionList.DropDownStyle = ComboBoxStyle.DropDownList;
            uxLedBuzzerControlSelectionList.Items.Clear();
            uxLedBuzzerControlSelectionList.Items.Add("0- Turn LED off");
            uxLedBuzzerControlSelectionList.Items.Add("1- Green on");
            uxLedBuzzerControlSelectionList.Items.Add("2- Green off");
            uxLedBuzzerControlSelectionList.Items.Add("3- Red on");
            uxLedBuzzerControlSelectionList.Items.Add("4- Red off");
            uxLedBuzzerControlSelectionList.Items.Add("5- 1 beep");
            uxLedBuzzerControlSelectionList.Items.Add("6- 3 beeps");
            uxLedBuzzerControlSelectionList.Items.Add("7- Green on with 1 beep");
            uxLedBuzzerControlSelectionList.Items.Add("8- Red on with 3 beeps");
            uxLedBuzzerControlSelectionList.Items.Add("9- Red, Green LED on");
            uxLedBuzzerControlSelectionList.Items.Add("C- Buzzer always beep");
            uxLedBuzzerControlSelectionList.Items.Add("D- Buzzer off");
            uxLedBuzzerControlSelectionList.SelectedIndex = 0;
        }
        #endregion

        #region UI Event Handlers
        private void Button1_Click(object sender, EventArgs e) => RefreshDevices();
        private void Button3_Click(object sender, EventArgs e) => OpenDoor();
        private void Button5_Click(object sender, EventArgs e) => ControlLedBuzzer();
        private void Button7_Click(object sender, EventArgs e) => RunBatchCommand(uxSelectedDeviceIp.Text);

        private void OpenDoor()
        {
            try
            {
                LogDebug($"Opening door on {uxSelectedDeviceIp.Text}");
                _er.OpenDoor(uxSelectedDeviceIp.Text,
                            (ushort)uxRemotePort.Value,
                            (byte)uxOpenDoorDuration.Value);
            }
            catch (Exception ex)
            {
                LogDebug($"Open door failed: {ex}");
                MessageBox.Show(ex.Message);
            }
        }

        private void ControlLedBuzzer()
        {
            try
            {
                byte function = GetSelectedFunction();
                LogDebug($"Controlling LED/Buzzer on {uxSelectedDeviceIp.Text} with function {function}");
                
                _er.ControlLedBuzzer(uxSelectedDeviceIp.Text,
                                   (ushort)uxRemotePort.Value,
                                   function);
            }
            catch (Exception ex)
            {
                LogDebug($"LED/Buzzer control failed: {ex}");
                MessageBox.Show(ex.Message);
            }
        }

        private byte GetSelectedFunction()
        {
            string selection = uxLedBuzzerControlSelectionList.Text.Substring(0, 1);
            switch (selection)
            {
                case "C": return (byte)('C' - '0');
                case "D": return (byte)('D' - '0');
                default: return (byte)uxLedBuzzerControlSelectionList.SelectedIndex;
            }
        }
        #endregion

        #region Missing Event Handlers (Added to fix errors)
        private void GroupBox3_Enter(object sender, EventArgs e) { /* Required but empty handler */ }
        private void GroupBox4_Enter(object sender, EventArgs e) { /* Required but empty handler */ }
        private void TextBox1_TextChanged(object sender, EventArgs e) { /* Required but empty handler */ }
        private void uxBatchCommandTimer_Tick(object sender, EventArgs e) => uxOpenDoorTimer_Tick(sender, e);
        private void uxLocalIps_SelectedIndexChanged(object sender, EventArgs e) { /* Required but empty handler */ }
        private void uxOpenDoorDuration_ValueChanged(object sender, EventArgs e) { /* Required but empty handler */ }
        private void uxSelectedDeviceIp_TextChanged(object sender, EventArgs e) { /* Required but empty handler */ }
        private void uxStartListen_Click(object sender, EventArgs e)
        {
            try
            {
                switch (uxStartListen.Text)
                {
                    case "▶ Start Listening":
                        _er.StartListen(uxLocalIps.Text, (ushort)uxListenPort.Value);
                        uxStartListen.Text = "⏹ Stop Listening";
                        LogDebug("Started listening for device data");
                        break;
                    case "⏹ Stop Listening":
                        _er.StopListen();
                        uxStartListen.Text = "▶ Start Listening";
                        LogDebug("Stopped listening for device data");
                        break;
                    default:
                        Debug.Assert(false);
                        break;
                }
            }
            catch (Exception ex)
            {
                LogDebug($"Listen control failed: {ex}");
                MessageBox.Show(ex.Message);
            }
        }
        #endregion

        #region Timer Events
        private void uxOpenDoorTimer_Tick(object sender, EventArgs e)
        {
            uxOpenDoorTimer.Enabled = false;
            try
            {
                string deviceIP = uxOpenDoorTimer.Tag as string;
                LogDebug($"Batch door open for {deviceIP}");
                
                _er.OpenDoor(deviceIP,
                            (ushort)uxRemotePort.Value,
                            (byte)uxOpenDoorDuration.Value,
                            false);
                
                uxControlLedBuzzerTimer.Tag = deviceIP;
                uxControlLedBuzzerTimer.Enabled = true;
            }
            catch (Exception ex)
            {
                LogDebug($"Batch door open failed: {ex}");
            }
        }

        private void uxControlLedBuzzerTimer_Tick(object sender, EventArgs e)
        {
            uxControlLedBuzzerTimer.Enabled = false;
            try
            {
                string deviceIP = uxControlLedBuzzerTimer.Tag as string;
                LogDebug($"Batch LED control for {deviceIP}");
                
                _er.ControlLedBuzzer(deviceIP,
                                   (ushort)uxRemotePort.Value,
                                   (byte)uxLedBuzzerControlSelectionList.SelectedIndex,
                                   false);
            }
            catch (Exception ex)
            {
                LogDebug($"Batch LED control failed: {ex}");
            }
        }
        #endregion

        #region Other Event Handlers
        private void uxDeviceListView_SelectedIndexChanged(object sender, EventArgs e)
        {
            if (uxDeviceListView.SelectedItems.Count > 0)
            {
                uxSelectedDeviceIp.Text = uxDeviceListView.SelectedItems[0].Text;
            }
        }

        private void uxEventDataListView_SelectedIndexChanged(object sender, EventArgs e)
        {
            if (uxEventDataListView.SelectedItems.Count > 0)
            {
                uxSelectedDeviceIp.Text = uxEventDataListView.SelectedItems[0].Text;
            }
        }

        private void Button4_Click(object sender, EventArgs e) => Application.Exit();
        private void Button2_Click_1(object sender, EventArgs e) => uxEventMessageListView.Items.Clear();
        private void Button6_Click_1(object sender, EventArgs e) => uxEventDataListView.Items.Clear();
        #endregion
    }
}